### ISSUE TITLE
token.WaitTimeout  There is an error missing.

### ISSUE THREAD

**Issue by yuelwish** [eclipse/paho.mqtt.golang#656]
paho.mqtt.golang version: 1.4.3

Regarding the MQTT connection, there is an issue where there is no error after the WaitTimeout times out, causing the overall logic to fail.

```
if token := mClient.Connect(); token.WaitTimeout(5*time.Second) && token.Error() != nil {
	return errors.Wrapf(token.Error(), "SerMQTT conn failed: %v", mClient.OptionsReader().ClientID())
}
```
Source code:

```
func (b *baseToken) WaitTimeout(d time.Duration) bool {
	timer := time.NewTimer(d)
	select {
	case <-b.complete:
		if !timer.Stop() {
			<-timer.C
		}
		return true
	case <-timer.C:
	}

	return false
}

```
When WaitTimeout times out, it returns false. At that time, token.Error() is nil.

As a result:

```
if false && nil != nil {
	// This block should be triggered as a timeout condition, but it cannot be reached.
}
```
In Go, there is a convention that when err == nil, it indicates that the current value is valid. When WaitTimeout times out, a connection timeout error should be set for baseToken.

**Comment by MattBrittan** [comment]
From [the docs](https://pkg.go.dev/github.com/eclipse/paho.mqtt.golang#Token):

>// WaitTimeout takes a time.Duration to wait for the flow associated with the
	// Token to complete, returns true if it returned before the timeout or
	// returns false if the timeout occurred. In the case of a timeout the Token
	// does not have an error set in case the caller wishes to wait again.

So I think this is pretty clear; `WaitTimeout` wait's until either the token completes or the timeout expires. Should the timeout expire then it returns false and the user should take whatever action is neccessary in this case. The `token` itself is not in an error state (the operation is still running and may complete successfully).

The issue here is the proliferation of 

```if token := client.Connect(); token.WaitTimeout(2*time.Second) && token.Error( d time.Duration) != nil {```

type operations. These will not do what the user expects should the operation timeout. There are quite a few of these in the example apps within this  repo and they need to be removed. I wonder if a helper func (something like `WaitOrError(Token, time.Duration) err`) would help.

>In Go, there is a convention that when err == nil, it indicates that the current value is valid.

This convention applies when a function has multiple return values, one of which is an `error`. This is not the case here.

Note that I'm not saying that this API is great, it's not, but it's also not going to change in any major way now as this will break too many users code. We have adopted a totally different interface with the v5 client (which I feel is much more usable and I've migrated most of my systems over to that).

**Comment by yuelwish** [comment]
But the test case told me that it was written like this.

This is where I learned this writing when I first came into contact with it.

paho.mqtt.golang@v1.4.3/client_test.go:65
```
// Try to connect using custom function, wait for 2 seconds, to pass MQTT first message
if token := client.Connect(); token.WaitTimeout(2*time.Second) && token.Error() != nil {
    t.Fatalf("%v", token.Error())
}
```

**Comment by MattBrittan** [comment]
>But the test case told me that it was written like this.

Yep - that's why I said that we need to remove all such references from the test apps (should have also specified tests!). Some bad code like this has crept in over the years (this never worked properly; `token.WaitTimeout` was only added three years ago whereas much of the rest of `token` is nine years old).

**Comment by MattBrittan** [comment]
Have updated the tests and added a function `WaitTokenTimeout` showing how to use `WaitTimeout` properly.

Interestingly the example you looked at (paho.mqtt.golang@v1.4.3/client_test.go:65) is actually expecting the call to `WaitTimeout` to return `false` (as the `Connect` operation should NOT complete in this test). So this was not a bug but it was confusing so I have tried to make it clear what the test is actually doing.
